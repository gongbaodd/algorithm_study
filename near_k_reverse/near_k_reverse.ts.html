<!DOCTYPE html>
<html><head><title>near_k_reverse/near_k_reverse.ts
 - Stryker report</title>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link rel="stylesheet" href="../strykerResources/bootstrap/css/bootstrap.min.css">
<link rel="stylesheet" href="../strykerResources/highlightjs/styles/default.css">
<link rel="stylesheet" href="../strykerResources/stryker/stryker.css"></head>
<body><img class="stryker-image" alt="Stryker" src="../strykerResources/stryker/stryker-80x80.png">
<div class="container-fluid"><div class="row"><div class="col-lg-12"><h1 class="display-4">near_k_reverse/near_k_reverse.ts
 
<small class="text-muted">- Stryker report</small></h1></div></div>
<ol class="breadcrumb"><li class="breadcrumb-item"><a href="../index.html">All files</a></li>
<li class="breadcrumb-item active">near_k_reverse/near_k_reverse.ts</li></ol>
<div class="row"><div class="col-lg-12"><div class="row"><div class="col-sm-11"><table class="table table-sm table-hover table-bordered table-no-top"><thead><tr><th style="width: 20%"><div><span>File / Directory</span></div></th>
<th colspan="2"><div><span>Mutation score</span></div></th>
<th class="rotate text-center" style="width: 50px"><div><span># Killed</span></div></th>
<th class="rotate text-center" style="width: 50px"><div><span># Survived</span></div></th>
<th class="rotate text-center" style="width: 50px"><div><span># Timeout</span></div></th>
<th class="rotate text-center" style="width: 50px"><div><span># No coverage</span></div></th>
<th class="rotate text-center" style="width: 50px"><div><span># Runtime errors</span></div></th>
<th class="rotate text-center" style="width: 50px"><div><span># Transpile errors</span></div></th>
<th class="rotate rotate-width-70 text-center" style="width: 70px"><div><span>Total detected</span></div></th>
<th class="rotate rotate-width-70 text-center" style="width: 70px"><div><span>Total undetected</span></div></th>
<th class="rotate rotate-width-70 text-center" style="width: 70px"><div><span>Total mutants</span></div></th></tr></thead>
<tbody><tr><td>near_k_reverse/near_k_reverse.ts</td>
<td><div class="progress"><div class="progress-bar bg-success" role="progressbar" aria-valuenow="84.62" aria-valuemin="0" aria-valuemax="100" style="width: 84.62%;">84.62
%</div></div></td>
<th class="text-center text-success" style="width: 10%">84.62</th>
<td class="text-center">21</td>
<td class="text-center">4</td>
<td class="text-center">1</td>
<td class="text-center">0</td>
<td class="text-center">13</td>
<td class="text-center">0</td>
<th class="text-center">22</th>
<th class="text-center">4</th>
<th class="text-center">39</th></tr>
</tbody></table></div></div>
<div class="row"><div class="col-lg-7"><div class="row legend"><form class="col-md-12" novalidate="novalidate">
<div class="form-check form-check-inline"><label class="form-check-label"><input class="form-check-input stryker-display" checked value="2" type="checkbox">
Survived
 
(4)</label></div>
<div class="form-check form-check-inline"><label class="form-check-label"><input class="form-check-input stryker-display" value="1" type="checkbox">
Killed
 
(21)</label></div>
<div class="form-check form-check-inline"><label class="form-check-label"><input class="form-check-input stryker-display" value="3" type="checkbox">
TimedOut
 
(1)</label></div>
<div class="form-check form-check-inline"><label class="form-check-label"><input class="form-check-input stryker-display" value="4" type="checkbox">
RuntimeError
 
(13)</label></div>

<a href="#" class="stryker-collapse-expand-all">Expand all</a></form></div>
<pre><code class="lang-javascript"><span class="bg-null">import { LNode } from &quot;../lib&quot;;

export function reverse(_head: LNode): LNode </span><a href="#" class="stryker-mutant-button" tabindex="0" title="Block" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="0"><span class="badge badge-secondary">0</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="0">{}</span><span class="stryker-original-code" data-mutant="0"><span class="bg-null">{
  const head = _head;
  if (</span><a href="#" class="stryker-mutant-button" tabindex="0" title="IfStatement" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="1"><span class="badge badge-secondary">1</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="1">true</span><a href="#" class="stryker-mutant-button" tabindex="0" title="IfStatement" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="2"><span class="badge badge-secondary">2</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="2">false</span><span class="stryker-original-code" data-mutant="1"><span class="stryker-original-code" data-mutant="2"><span class="bg-null">head.next </span><a href="#" class="stryker-mutant-button" tabindex="0" title="BinaryExpression" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="3"><span class="badge badge-secondary">3</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="3">!==</span><span class="stryker-original-code" data-mutant="3"><span class="bg-null">===</span></span><span class="bg-null"> null</span></span></span><span class="bg-null">) </span><a href="#" class="stryker-mutant-button" tabindex="0" title="Block" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="4"><span class="badge badge-secondary">4</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="4">{}</span><span class="stryker-original-code" data-mutant="4"><span class="bg-null">{
    return head;
  }</span></span><span class="bg-null">

  const newHead = reverse(head.next);
  </span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="5"><span class="badge badge-success">5</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="5">false</span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="6"><span class="badge badge-success">6</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="6">true</span><span class="stryker-original-code" data-mutant="5"><span class="stryker-original-code" data-mutant="6"><span class="bg-success-light">head.next.next = head</span></span></span><span class="bg-null">;
  </span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Survived" data-mutant-status-annotation="danger" data-mutant-status="2" data-mutant="7"><span class="badge badge-danger">7</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="7">false</span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Survived" data-mutant-status-annotation="danger" data-mutant-status="2" data-mutant="8"><span class="badge badge-danger">8</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="8">true</span><span class="stryker-original-code" data-mutant="7"><span class="stryker-original-code" data-mutant="8"><span class="bg-danger-light">head.next = null</span></span></span><span class="bg-null">;

  return newHead;
}</span></span><span class="bg-null">

export function nearkReverse(head: LNode, k: number): LNode </span><a href="#" class="stryker-mutant-button" tabindex="0" title="Block" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="9"><span class="badge badge-secondary">9</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="9">{}</span><span class="stryker-original-code" data-mutant="9"><span class="bg-null">{
  if (</span><a href="#" class="stryker-mutant-button" tabindex="0" title="IfStatement" data-content="status: Survived" data-mutant-status-annotation="danger" data-mutant-status="2" data-mutant="10"><span class="badge badge-danger">10</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="10">false</span><a href="#" class="stryker-mutant-button" tabindex="0" title="IfStatement" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="11"><span class="badge badge-secondary">11</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="11">true</span><span class="stryker-original-code" data-mutant="10"><span class="stryker-original-code" data-mutant="11"><span class="bg-danger-light">head.next </span><a href="#" class="stryker-mutant-button" tabindex="0" title="BinaryExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="12"><span class="badge badge-success">12</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="12">!==</span><span class="stryker-original-code" data-mutant="12"><span class="bg-danger-light">===</span></span><span class="bg-danger-light"> null</span></span></span><span class="bg-null">) </span><a href="#" class="stryker-mutant-button" tabindex="0" title="Block" data-content="status: Survived" data-mutant-status-annotation="danger" data-mutant-status="2" data-mutant="13"><span class="badge badge-danger">13</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="13">{}</span><span class="stryker-original-code" data-mutant="13"><span class="bg-danger-light">{
    return head;
  }</span></span><span class="bg-null">

  let pre = head;
  let current = pre.next;

  while (</span><a href="#" class="stryker-mutant-button" tabindex="0" title="WhileStatement" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="14"><span class="badge badge-secondary">14</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="14">false</span><span class="stryker-original-code" data-mutant="14"><span class="bg-null">current</span></span><span class="bg-null">) </span><a href="#" class="stryker-mutant-button" tabindex="0" title="Block" data-content="status: TimedOut" data-mutant-status-annotation="warning" data-mutant-status="3" data-mutant="15"><span class="badge badge-warning">15</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="15">{}</span><span class="stryker-original-code" data-mutant="15"><span class="bg-warning-light">{
    let next: LNode | null = current;
    for (let i = 0; </span><a href="#" class="stryker-mutant-button" tabindex="0" title="ForStatement" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="16"><span class="badge badge-success">16</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="16">false</span><span class="stryker-original-code" data-mutant="16"><span class="bg-warning-light">i </span><a href="#" class="stryker-mutant-button" tabindex="0" title="BinaryExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="17"><span class="badge badge-success">17</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="17">&lt;=</span><a href="#" class="stryker-mutant-button" tabindex="0" title="BinaryExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="18"><span class="badge badge-success">18</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="18">&gt;=</span><span class="stryker-original-code" data-mutant="17"><span class="stryker-original-code" data-mutant="18"><span class="bg-warning-light">&lt;</span></span></span><span class="bg-warning-light"> k </span><a href="#" class="stryker-mutant-button" tabindex="0" title="BinaryExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="19"><span class="badge badge-success">19</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="19">+</span><span class="stryker-original-code" data-mutant="19"><span class="bg-warning-light">-</span></span><span class="bg-warning-light"> 1</span></span><span class="bg-warning-light">; i += 1) </span><a href="#" class="stryker-mutant-button" tabindex="0" title="Block" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="20"><span class="badge badge-success">20</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="20">{}</span><span class="stryker-original-code" data-mutant="20"><span class="bg-warning-light">{
      </span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="21"><span class="badge badge-success">21</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="21">false</span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="22"><span class="badge badge-success">22</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="22">true</span><span class="stryker-original-code" data-mutant="21"><span class="stryker-original-code" data-mutant="22"><span class="bg-warning-light">next = </span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="23"><span class="badge badge-secondary">23</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="23">false</span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="24"><span class="badge badge-secondary">24</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="24">true</span><span class="stryker-original-code" data-mutant="23"><span class="stryker-original-code" data-mutant="24"><span class="bg-warning-light">next?.next ?? null</span></span></span></span></span><span class="bg-warning-light">;
      if (</span><a href="#" class="stryker-mutant-button" tabindex="0" title="IfStatement" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="25"><span class="badge badge-success">25</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="25">true</span><a href="#" class="stryker-mutant-button" tabindex="0" title="IfStatement" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="26"><span class="badge badge-secondary">26</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="26">false</span><span class="stryker-original-code" data-mutant="25"><span class="stryker-original-code" data-mutant="26"><span class="bg-warning-light">next </span><a href="#" class="stryker-mutant-button" tabindex="0" title="BinaryExpression" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="27"><span class="badge badge-secondary">27</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="27">!==</span><span class="stryker-original-code" data-mutant="27"><span class="bg-warning-light">===</span></span><span class="bg-warning-light"> null</span></span></span><span class="bg-warning-light">) </span><a href="#" class="stryker-mutant-button" tabindex="0" title="Block" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="28"><span class="badge badge-secondary">28</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="28">{}</span><span class="stryker-original-code" data-mutant="28"><span class="bg-warning-light">{
        </span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="29"><span class="badge badge-success">29</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="29">false</span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="30"><span class="badge badge-success">30</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="30">true</span><span class="stryker-original-code" data-mutant="29"><span class="stryker-original-code" data-mutant="30"><span class="bg-warning-light">pre.next = current</span></span></span><span class="bg-warning-light">;
        return head;
      }</span></span><span class="bg-warning-light">
    }</span></span><span class="bg-warning-light">
    const nNext = next.next;
    </span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="31"><span class="badge badge-success">31</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="31">false</span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="32"><span class="badge badge-success">32</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="32">true</span><span class="stryker-original-code" data-mutant="31"><span class="stryker-original-code" data-mutant="32"><span class="bg-warning-light">next.next = null</span></span></span><span class="bg-warning-light">;

    </span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="33"><span class="badge badge-success">33</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="33">false</span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="34"><span class="badge badge-success">34</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="34">true</span><span class="stryker-original-code" data-mutant="33"><span class="stryker-original-code" data-mutant="34"><span class="bg-warning-light">pre.next = reverse(current)</span></span></span><span class="bg-warning-light">;
    </span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="35"><span class="badge badge-success">35</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="35">false</span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="36"><span class="badge badge-success">36</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="36">true</span><span class="stryker-original-code" data-mutant="35"><span class="stryker-original-code" data-mutant="36"><span class="bg-warning-light">pre = current</span></span></span><span class="bg-warning-light">;
    </span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="37"><span class="badge badge-success">37</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="37">false</span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="38"><span class="badge badge-success">38</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="38">true</span><span class="stryker-original-code" data-mutant="37"><span class="stryker-original-code" data-mutant="38"><span class="bg-warning-light">current = nNext</span></span></span><span class="bg-warning-light">;
  }</span></span><span class="bg-null">

  return head;
}</span></span><span class="bg-null">
</code></pre></div>
<div class="col-lg-5"><table class="table table-sm table-hover mutant-table"><thead><tr><th>#</th>
<th>Mutator</th>
<th>State</th>
<th>Location</th>
<th>Original</th>
<th>Replacement</th></tr></thead>
<tbody><tr><th>0</th>
<td>Block</td>
<td>RuntimeError</td>
<td>2
:
45</td>
<td><code>{
   ...  ;
}</code></td>
<td><code>{}</code></td></tr>
<tr><th>1</th>
<td>IfStatement</td>
<td>RuntimeError</td>
<td>4
:
6</td>
<td><code>.     ===</code></td>
<td><code></code></td></tr>
<tr><th>2</th>
<td>IfStatement</td>
<td>RuntimeError</td>
<td>4
:
6</td>
<td><code>.     ===</code></td>
<td><code></code></td></tr>
<tr><th>3</th>
<td>BinaryExpression</td>
<td>RuntimeError</td>
<td>4
:
16</td>
<td><code>===</code></td>
<td><code>!==</code></td></tr>
<tr><th>4</th>
<td>Block</td>
<td>RuntimeError</td>
<td>4
:
26</td>
<td><code>{
   ...;
  }</code></td>
<td><code>{}</code></td></tr>
<tr><th>5</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>9
:
2</td>
<td><code>.    .     =</code></td>
<td><code></code></td></tr>
<tr><th>6</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>9
:
2</td>
<td><code>.    .     =</code></td>
<td><code></code></td></tr>
<tr><th>7</th>
<td>ConditionalExpression</td>
<td>Survived</td>
<td>10
:
2</td>
<td><code>.     =</code></td>
<td><code></code></td></tr>
<tr><th>8</th>
<td>ConditionalExpression</td>
<td>Survived</td>
<td>10
:
2</td>
<td><code>.     =</code></td>
<td><code></code></td></tr>
<tr><th>9</th>
<td>Block</td>
<td>RuntimeError</td>
<td>15
:
60</td>
<td><code>{
   ...  ;
}</code></td>
<td><code>{}</code></td></tr>
<tr><th>10</th>
<td>IfStatement</td>
<td>Survived</td>
<td>16
:
6</td>
<td><code>.     ===</code></td>
<td><code></code></td></tr>
<tr><th>11</th>
<td>IfStatement</td>
<td>RuntimeError</td>
<td>16
:
6</td>
<td><code>.     ===</code></td>
<td><code></code></td></tr>
<tr><th>12</th>
<td>BinaryExpression</td>
<td>Killed</td>
<td>16
:
16</td>
<td><code>===</code></td>
<td><code>!==</code></td></tr>
<tr><th>13</th>
<td>Block</td>
<td>Survived</td>
<td>16
:
26</td>
<td><code>{
   ...;
  }</code></td>
<td><code>{}</code></td></tr>
<tr><th>14</th>
<td>WhileStatement</td>
<td>RuntimeError</td>
<td>23
:
9</td>
<td><code></code></td>
<td><code></code></td></tr>
<tr><th>15</th>
<td>Block</td>
<td>TimedOut</td>
<td>23
:
18</td>
<td><code>{
   ...;
  }</code></td>
<td><code>{}</code></td></tr>
<tr><th>16</th>
<td>ForStatement</td>
<td>Killed</td>
<td>25
:
20</td>
<td><code><   -</code></td>
<td><code></code></td></tr>
<tr><th>17</th>
<td>BinaryExpression</td>
<td>Killed</td>
<td>25
:
22</td>
<td><code><</code></td>
<td><code><=</code></td></tr>
<tr><th>18</th>
<td>BinaryExpression</td>
<td>Killed</td>
<td>25
:
22</td>
<td><code><</code></td>
<td><code>>=</code></td></tr>
<tr><th>19</th>
<td>BinaryExpression</td>
<td>Killed</td>
<td>25
:
26</td>
<td><code>-</code></td>
<td><code>+</code></td></tr>
<tr><th>20</th>
<td>Block</td>
<td>Killed</td>
<td>25
:
39</td>
<td><code>{
   ...    }</code></td>
<td><code>{}</code></td></tr>
<tr><th>21</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>26
:
6</td>
<td><code>=     ?.     ??</code></td>
<td><code></code></td></tr>
<tr><th>22</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>26
:
6</td>
<td><code>=     ?.     ??</code></td>
<td><code></code></td></tr>
<tr><th>23</th>
<td>ConditionalExpression</td>
<td>RuntimeError</td>
<td>26
:
13</td>
<td><code>?.     ??</code></td>
<td><code></code></td></tr>
<tr><th>24</th>
<td>ConditionalExpression</td>
<td>RuntimeError</td>
<td>26
:
13</td>
<td><code>?.     ??</code></td>
<td><code></code></td></tr>
<tr><th>25</th>
<td>IfStatement</td>
<td>Killed</td>
<td>27
:
10</td>
<td><code>===</code></td>
<td><code></code></td></tr>
<tr><th>26</th>
<td>IfStatement</td>
<td>RuntimeError</td>
<td>27
:
10</td>
<td><code>===</code></td>
<td><code></code></td></tr>
<tr><th>27</th>
<td>BinaryExpression</td>
<td>RuntimeError</td>
<td>27
:
15</td>
<td><code>===</code></td>
<td><code>!==</code></td></tr>
<tr><th>28</th>
<td>Block</td>
<td>RuntimeError</td>
<td>27
:
25</td>
<td><code>{
   ...    }</code></td>
<td><code>{}</code></td></tr>
<tr><th>29</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>28
:
8</td>
<td><code>.     =</code></td>
<td><code></code></td></tr>
<tr><th>30</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>28
:
8</td>
<td><code>.     =</code></td>
<td><code></code></td></tr>
<tr><th>31</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>33
:
4</td>
<td><code>.     =</code></td>
<td><code></code></td></tr>
<tr><th>32</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>33
:
4</td>
<td><code>.     =</code></td>
<td><code></code></td></tr>
<tr><th>33</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>35
:
4</td>
<td><code>.    ...    )</code></td>
<td><code></code></td></tr>
<tr><th>34</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>35
:
4</td>
<td><code>.    ...    )</code></td>
<td><code></code></td></tr>
<tr><th>35</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>36
:
4</td>
<td><code>=</code></td>
<td><code></code></td></tr>
<tr><th>36</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>36
:
4</td>
<td><code>=</code></td>
<td><code></code></td></tr>
<tr><th>37</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>37
:
4</td>
<td><code>=</code></td>
<td><code></code></td></tr>
<tr><th>38</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>37
:
4</td>
<td><code>=</code></td>
<td><code></code></td></tr></tbody></table></div></div></div></div>
<footer>Generated with stryker-html-reporter generator. Visit the 
<a href="https://stryker-mutator.io" target="_blank">Stryker website</a></footer></div>
<script src="../strykerResources/popper.js/dist/umd/popper.min.js" defer="defer"></script>
<script src="../strykerResources/jquery/dist/jquery.slim.min.js" defer="defer"></script>
<script src="../strykerResources/bootstrap/js/bootstrap.min.js" defer="defer"></script>
<script src="../strykerResources/stryker.js" defer="defer"></script></body></html>