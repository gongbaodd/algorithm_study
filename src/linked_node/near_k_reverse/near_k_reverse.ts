import { LNode } from "../lib";

export function reverse(_head: LNode): LNode {
  const head = _head;
  if (head.next === null) {
    return head;
  }

  const newHead = reverse(head.next);
  head.next.next = head;
  head.next = null;

  return newHead;
}

export function nearkReverse(head: LNode, k: number): LNode {
  if (head.next === null) {
    return head;
  }

  let pre = head;
  let current = pre.next;

  while (current) {
    let next: LNode | null = current;
    for (let i = 0; i < k - 1; i += 1) {
      next = next?.next ?? null;
      if (next === null) {
        pre.next = current;
        return head;
      }
    }
    const nNext = next.next;
    next.next = null;

    pre.next = reverse(current);
    pre = current;
    current = nNext;
  }

  return head;
}
