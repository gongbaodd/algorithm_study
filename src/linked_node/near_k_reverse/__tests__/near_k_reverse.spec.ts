import { nearkReverse } from "../near_k_reverse";
import { init } from "../../lib";

describe("near k reverse", () => {
  describe("k === 2", () => {
    it("should return 2,1,4,3,6,5,7", () => {
      expect(nearkReverse(init([1, 2, 3, 4, 5, 6, 7]), 2)).toMatchSnapshot();
    });
  });

  describe("k === 3", () => {
    it("should return 3,2,1,6,5,4,7", () => {
      expect(nearkReverse(init([1, 2, 3, 4, 5, 6, 7]), 3)).toMatchSnapshot();
    });
  });
});
