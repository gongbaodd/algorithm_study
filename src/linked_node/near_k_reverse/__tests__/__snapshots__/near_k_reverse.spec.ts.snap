// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`near k reverse k === 2 should return 2,1,4,3,6,5,7 1`] = `
LNode {
  "data": 0,
  "next": LNode {
    "data": 2,
    "next": LNode {
      "data": 1,
      "next": LNode {
        "data": 4,
        "next": LNode {
          "data": 3,
          "next": LNode {
            "data": 6,
            "next": LNode {
              "data": 5,
              "next": LNode {
                "data": 7,
                "next": null,
              },
            },
          },
        },
      },
    },
  },
}
`;

exports[`near k reverse k === 3 should return 3,2,1,6,5,4,7 1`] = `
LNode {
  "data": 0,
  "next": LNode {
    "data": 3,
    "next": LNode {
      "data": 2,
      "next": LNode {
        "data": 1,
        "next": LNode {
          "data": 6,
          "next": LNode {
            "data": 5,
            "next": LNode {
              "data": 4,
              "next": LNode {
                "data": 7,
                "next": null,
              },
            },
          },
        },
      },
    },
  },
}
`;
